### PR TITLE
Fix BFloat16 related failures and warnings clean up

### PR DESCRIPTION
**Pull request by jerryzj**
Request review: @aswaterman @nibrunieAtSi5

### CODE REVIEW

**Review by nibrunieAtSi5 — COMMENTED** [review]
To make sure this is better tested than my PR, could you list which target(s) you have built and which test you ran ?

**Comment by jerryzj** [comment]
Sure, I will test it on RISCV64(QEMU)and X86(internal machine)

**Comment by jerryzj** [comment]
TestFloat results (`./testfloat -all1 -level 1`)

* Test configurations
  * x86, intel Xeon(R) Platinum 8460Y+
    * Compiler: gcc 13.2.0
  * x86, AMD Ryzen 9 3950X
    * Compiler: gcc 12.3.0
  * RISCV64, rv64g_zfh_zfhmin_zfbfmin0p8
    * Compiler: SiFive in-house gcc 13.2.0 

* Without BFloat16 ([06b20075dd3c1a5d0dd007a93643282832221612](https://github.com/ucb-bar/berkeley-testfloat-3/commit/06b20075dd3c1a5d0dd007a93643282832221612))
  * x86: Some tests failed, tested on intel and AMD  have the same failures

```
Errors found in f32_to_ui32_rx_minMag:
-82.40000E  => FFFFFFF4 ....x  expected FFFFFFFF v....
-85.7FFD7F  => FFFFFF81 ....x  expected FFFFFFFF v....
-7F.21FFFF  => FFFFFFFF ....x  expected FFFFFFFF v....
+BD.3D9023  => 00000000 .....  expected FFFFFFFF v....
-82.7FBFEE  => FFFFFFF1 ....x  expected FFFFFFFF v....
(skip)
Errors found in f32_to_ui64_rx_minMag:
-9E.01FFFC  => FFFFFFFF7E000400 .....  expected FFFFFFFFFFFFFFFF v....
+FF.4C1A43  => 8000000000000000 v....  expected FFFFFFFFFFFFFFFF v....
-BC.001FFD  => DFF800C000000000 .....  expected FFFFFFFFFFFFFFFF v....
-8D.6DA13E  => FFFFFFFFFFFF8930 ....x  expected FFFFFFFFFFFFFFFF v....
(skip)
Errors found in f64_to_ui32_rx_minMag:
+41F.07FFFFFFFFDFE  => 07FFFFFF ....x  expected FFFFFFFF v....
-439.FFFFFFFE00800  => 07FE0000 .....  expected FFFFFFFF v....
-403.C0000000000FF  => FFFFFFE4 ....x  expected FFFFFFFF v....
+41F.000800000001E  => 00080000 ....x  expected FFFFFFFF v....
(skip)
Errors found in f64_to_ui64_rx_minMag:
-41E.4F01909446C11  => FFFFFFFF587F37B6 ....x  expected FFFFFFFFFFFFFFFF v....
-430.FFFFC20000000  => FFFC00007C000000 .....  expected FFFFFFFFFFFFFFFF v....
-50D.FFBFFFFC00000  => 8000000000000000 v....  expected FFFFFFFFFFFFFFFF v....
-41D.F13481AFD611C  => FFFFFFFF83B2DF95 ....x  expected FFFFFFFFFFFFFFFF v....
(skip)
Errors found in extF80_to_ui32_rx_minMag:
-403C.FE000007FFFFFFFE  => 00000001 ....x  expected FFFFFFFF v....
-407E.80800001FFFFFFFF  => 00000000 v....  expected FFFFFFFF v....
-401E.FFFFFFFBFFFF8000  => 00000005 ....x  expected FFFFFFFF v....
-4000.DB4EAB1124AAB57E  => FFFFFFFD ....x  expected FFFFFFFF v....
(skip)
Errors found in extF80_to_ui64_rx_minMag:
+6B7F.D1658B972B208E2B
	=> 0000000000000000 v...x  expected FFFFFFFFFFFFFFFF v....
-403E.A559E21659DF2CC4
	=> 8000000000000000 v....  expected FFFFFFFFFFFFFFFF v....
-403D.8000000000005FFF
	=> BFFFFFFFFFFFD001 ....x  expected FFFFFFFFFFFFFFFF v....
-407F.80000000000FFFFF
	=> 8000000000000000 v....  expected FFFFFFFFFFFFFFFF v....
(skip)
Errors found in extF80_to_i32_rx_minMag:
+401E.8002001FFFFFFFFF  => 8002001F ....x  expected 80000000 v....
-407E.8001000000001FFF  => 00000000 v....  expected 80000000 v....
-4036.808000000000001E  => 00000000 ....x  expected 80000000 v....
-4047.FFFFFC0000000040  => 00000000 v....  expected 80000000 v....
(skip)
```
  * RISCV64: N/A
* With BFlaot16 ([105dfcd63197e262eebf15710ca89dc472dc6a04](https://github.com/ucb-bar/berkeley-testfloat-3/commit/7d9e6fcda0677a51631b121070330d86375da261))
  * x86: SEGV

```
Testing extF80_to_f16, rounding max.
Testing extF80_roundToInt_r_near_maxMag.
912 tests total.
[1]    1681933 segmentation fault  ./testfloat -all1 -level 1
```
  * RISCV64: Build failed
* This patch
  * x86: Same x86 failures as shown above
  * RISCV64: PASS (fp32/fp64/i32/i64/ui32/ui64 tests only)

File: source/genCases_bf16.c [hunk 1]
```diff
@@ -350,6 +352,8 @@ static bfloat16_t bf16Random( void )
         return bf16RandomQInfP3();
      case 7:
         return bf16RandomQInfPInf();
+     default:
+        __builtin_unreachable();
```
Comment (aswaterman) [inline source/genCases_bf16.c:356]: `__builtin_unreachable` is not standard C.  Please find a different way of expressing this that is standards-compliant.

Comment (aswaterman) [reply source/genCases_bf16.c:356]: You might be able to use something like `(void) 0;`.

File: source/timesoftfloat.c [hunk 2]
```diff
@@ -2876,38 +2876,38 @@ enum { numInputs_extF80 = 32 };
 static
 const union { struct extFloat80M s; extFloat80_t f; }
     inputs_extF80[numInputs_extF80] = {
-        extF80Const( 0xC03F, 0xA9BE15A19C1E8B62 ),
```
Comment (aswaterman) [inline source/timesoftfloat.c:2879]: This is clearly unrelated to BF16.  What problem is being fixed?  Is the solution C99-compliant?

**Comment by aswaterman** [comment]
@jerryz123 Can you provide a review, since this is a substantive change to a ucb-bar repo?

File: source/genCases_bf16.c (hunk 1 above)
Comment (jerryzj) [reply source/genCases_bf16.c:356]: Could we just use `abort()`? It's defined in `stdlib.h`

**Comment by jerryz123** [comment]
> @jerryz123 Can you provide a review, since this is a substantive change to a ucb-bar repo?

Sure. I'll get to this tomorrow

File: source/timesoftfloat.c (hunk 2 above)
Comment (jerryzj) [reply source/timesoftfloat.c:2879]: It repairs compiler warning like `warning: missing braces around initializer`

File: source/genCases_bf16.c (hunk 1 above)
Comment (aswaterman) [reply source/genCases_bf16.c:356]: Yeah, that is better than my suggestion.

File: source/timesoftfloat.c (hunk 2 above)
Comment (aswaterman) [reply source/timesoftfloat.c:2879]: Sounds good if it doesn’t break anything.

File: source/genCases_bf16.c (hunk 1 above)
Comment (jerryzj) [reply source/genCases_bf16.c]: Updated.

**Comment by jerryz123** [comment]
> This patch
> 
> * x86: Same x86 failures as shown above

To be clear, this does fix the segfaults,

**Review by jerryz123 — APPROVED** [review]
I assume the warnings only appear in gcc>=13, as I don't see anything on 12.

**Comment by nibrunieAtSi5** [comment]
I am still a bit worried about the test failures on x86 before BFloat16 introduction. Are those known issues ?

**Comment by jerryz123** [comment]
> I am still a bit worried about the test failures on x86 before BFloat16 introduction. Are those known issues ?

Yeah... those failures appear for me when using testfloat-3e and softfloat-3e... I think we'd have to ask John Hauser

**Review by aswaterman — APPROVED** [review]

**Comment by aswaterman** [comment]
I'll merge since the x86 failures are unrelated.  (I'll try to dig into the failures separately.)

**Comment by aswaterman** [comment]
Analysis so far:

GCC on x86 takes advantage of the fact that the results of out-of-bounds float-to-int conversions produce unspecified results.  Considering the first failure, when converting from -12.00001f to uint32_t, it produces (uint32_t)-12, which is weird but not invalid.  SoftFloat for x86 returns (uint32_t)-1 for anything < 0, regardless of magnitude.  TestFloat wants these results to exactly match, which, well...

**Comment by aswaterman** [comment]
My mistake: TestFloat is indeed skipping the data value comparison for the invalid case.  That's good.  It is actually complaining because the Invalid Operation flag wasn't set for an out-of-bounds conversion.

Other people have complained about this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=27682

I would think that for `__STDC_IEC_559__` being defined and with `#pragma STDC FENV_ACCESS ON`, this should do the right thing.  So it seems to be a bug in the C implementation.

**Pull request merged**